src/sentry/unmerge.py

```diff
@@ -89,25 +98,117 @@ def get_unmerge_key(
     def primary_hashes_to_lock(self) -> Collection[str]:
         return self.fingerprints
 
-    def start_snuba_replacement(self, project: Project, source_id: int, destination_id: int) -> Any:
+    def start_snuba_replacement(
+        self, project: Project, source_id: int, unmerge_key: str, destination_id: int
+    ) -> Any:
         return eventstream.start_unmerge(project.id, self.fingerprints, source_id, destination_id)
 
     def stop_snuba_replacement(self, eventstream_state: Any) -> None:
         if eventstream_state:
             eventstream.end_unmerge(eventstream_state)
 
     def run_postgres_replacement(
-        self, project: Project, destination_id: int, locked_primary_hashes: Collection[str]
+        self,
+        project: Project,
+        unmerge_key: str,
+        destination_id: int,
+        locked_primary_hashes: Collection[str],
     ) -> None:
         # Move the group hashes to the destination.
         GroupHash.objects.filter(project_id=project.id, hash__in=locked_primary_hashes).update(
             group=destination_id
         )
 
-    def get_activity_args(self) -> Mapping[str, Any]:
+    def get_activity_args(self, unmerge_key: str) -> Mapping[str, Any]:
+        assert unmerge_key == _DEFAULT_UNMERGE_KEY
         return {"fingerprints": self.fingerprints}
 
 
+@dataclass(frozen=True)
+class HierarchicalUnmergeReplacement(UnmergeReplacement):
+    """
+    Split up one issue by a particular entry in `hierarchical_hashes` into many
+    issues.
+
+    In contrast to `PrimaryHashUnmergeReplacement`, this produces multiple new
+    groups instead of just one.
+    """
+
+    primary_hash: str
+    filter_hierarchical_hash: str
+    filter_level: int
+    new_level: int
+    assume_source_emptied: bool
+    reset_hashes: Sequence[str]
+
+    def get_unmerge_key(
+        self, event: Event, locked_primary_hashes: Collection[str]
+    ) -> Optional[str]:
+        if event.get_primary_hash() != self.primary_hash:
+            return None
+
+        hierarchical_hashes = event.data.get("hierarchical_hashes")
+
+        if not hierarchical_hashes:
+            return None
+
+        try:
+            if hierarchical_hashes[self.filter_level] != self.filter_hierarchical_hash:
+                return None
+        except IndexError:
+            return None
+
+        try:
+            return hierarchical_hashes[self.new_level]
+        except IndexError:
+            return hierarchical_hashes[-1]
+
+    @property
+    def primary_hashes_to_lock(self) -> Collection[str]:
```

Comment:
Not sure what we would lock here. We cannot lock primary hash itself as in the case of hierarchical grouping, there may be concurrent unmerges on different groups that contain the same primary_hash (this is not the case for regular unmerge)

For now just allow races, we can figure out a better synchronization mechanism later